### PR TITLE
Implement RMSNorm layer with proper forward and backward pass

### PR DESCRIPTION
**Pull request by niket-agarwal**
- Add forwarding for training mode which internally calls incremental_forwarding
- Add calcDerivative and calcGradient implementation for backpropagation support
- Add nnlayergolden data file for RMSNorm_trainable inside unittest_layers.tar.gz
- Update unit tests to validate the implementation
- Move RMSNorm from Applications/CausalLM to nntrainer/layers

## Files Modified
`Applications/CausalLM/jni/Android.mk`
`Applications/CausalLM/layers/meson.build`
`Applications/CausalLM/layers/rms_norm.cpp`
`Applications/CausalLM/layers/rms_norm.h`
`Applications/CausalLM/meson.build`
`Applications/CausalLM/models/transformer.cpp`
`debian/nntrainer-dev.install`
`nntrainer/app_context.cpp`
`nntrainer/layers/meson.build`
`nntrainer/layers/rms_norm.cpp`
`nntrainer/layers/rms_norm.h`
`packaging/nntrainer.spec`
`packaging/unittest_layers.tar.gz`
`test/input_gen/gen_layer_tests.py`
`test/unittest/layers/meson.build`
`test/unittest/layers/unittest_layers_rms_norm.cpp`

Signed-off-by: Niket Agarwal <niket.a@samsung.com>

### CODE REVIEW

File: test/input_gen/gen_layer_tests.py [hunk 1]
```diff
@@ -950,10 +950,38 @@ def call(self, inputs):
...
+                initializer=tf.keras.initializers.Ones(),
+                trainable=True,
+                name='gamma'
+            )
+            super(RMSNorm_trainable, self).build(input_shape)
+
+        def call(self, inputs):
+            # Compute the mean of the squares of the inputs along the last dimension
+            mean_square = tf.reduce_mean(tf.square(inputs), axis=[-1], keepdims=True)
+            print(mean_square)
```
Comment (EunjuYang) [inline test/input_gen/gen_layer_tests.py]: Please drop these debug `print`s before mering - they'll spam the golden-generator output.
```suggestion
```

File: nntrainer/layers/rms_norm.cpp [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+ * @date   19 July 2023
+ * @brief  Implementation of RMS normalization function
+ * @see    https://github.com/nntrainer/nntrainer
+ * @author Seungbaek Hong <sb92.hong@samsung.com>
+ * @bug    No known bugs except for NYI items
+ *
+ */
+
+#include <cmath>
+#include <iostream>
```
Comment (EunjuYang) [inline nntrainer/layers/rms_norm.cpp:15]: Either mover `#include <iostream>` under `#ifdef DEBUG` or drop it - it's unused in release builds.

File: nntrainer/layers/rms_norm.h [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+   * @brief Construct a new RMS normalization layer object
+   *
+   */
+  RMSNormLayer() : Layer(), wt_idx({0}) {}
+
+  /**
+   * @brief Destroy the RMS normalization layer object
+   *
+   */
+  ~RMSNormLayer() {}
```
Comment (EunjuYang) [inline nntrainer/layers/rms_norm.h:49]: nit: `~RMSNormLayer() = default;`

**Review by EunjuYang — COMMENTED** [review]
LGTM with nits.

**Review by baek2sm — APPROVED** [review]
LGTM

File: nntrainer/layers/rms_norm.cpp [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    Tensor out_step =
+      out.getSharedDataTensor(out_step_dim, b * out_dim.getFeatureLen(), true);
+
+    if (in_step.getDataType() == ml::train::TensorDim::DataType::FP32) {
+      auto t = in_step.multiply(in_step).average(3).add(epsilon);
+      t.inv_sqrt_i();
+      in_step.multiply(t, out_step);
+    } else {
+      throw std::invalid_argument(
+        "Error: not yet implemented for this data type");
```
Comment (Jungwon-Lee) [inline nntrainer/layers/rms_norm.cpp:75]: minor: "RMSNorm incremental_forwarding: FP16 is not implemented yet"

File: nntrainer/layers/rms_norm.cpp [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+      TensorDim::TensorType(in.getDim().getFormat(), in.getDataType()));
+    Tensor inv_rms(inv_rms_dim, true);
+
+    // Compute inv_rms: inv_rms = 1 / sqrt(mean(x^2) + epsilon)
+    // Use a temporary tensor for computation
+    Tensor temp(in.getDim(), true);
+    in.multiply(in, temp);    // temp = x^2
+    temp.average(3, inv_rms); // inv_rms = mean(x^2) along width
+    inv_rms.add_i(epsilon);   // inv_rms = mean(x^2) + epsilon
+    inv_rms.inv_sqrt_i();     // inv_rms = 1 / sqrt(mean(x^2) + epsilon)
```
Comment (Jungwon-Lee) [inline nntrainer/layers/rms_norm.cpp:185]: minor: 

In the current implementation, `inv_rms` is computed twice:

1. During `incremental_forwarding()` (forward pass)
2. Again in `calcDerivative()` and `calcGradient()` (backward pass)

You can save and reuse the inv_rms like LayerNorm Implementation
(https://github.com/nntrainer/nntrainer/blob/b719613fecf2caa14e6364c9f48a8c9337bb9afb/nntrainer/layers/layer_normalization_layer.cpp#L249)

**Review by Jungwon-Lee — APPROVED** [review]

**Comment by baek2sm** [comment]
**Final Verdict: ✅ Approve**

---

### 1. Core Summary of Implemented Feature

Promotes `RMSNormLayer` from a CausalLM-local plugin into a core `nntrainer` layer with a complete, training-capable forward + backward pass.

**Included features**
- New core `nntrainer::RMSNormLayer` (forward, `calcDerivative`, `calcGradient`), registered in `AppContext`.
- CausalLM `transformer.cpp` switched to the core layer; the local copy and its registration removed.
- Golden-data unit test + Python reference; packaging/Debian manifests updated.

---

### 2. Related Issues

- **P3** Rebase conflict in `Applications/CausalLM/layers/meson.build` resolved (removed the `causallm_rms_norm` stanzas, preserved newer upstream entries; dropped a stray `openmp_dep`).
- **P3** Fixed an include-guard collision with `reshaped_rms_norm.h` (`__NNTRAINER_RMS_NORM_LAYER_H__`) and a `calcDerivative` scratch-buffer aliasing risk (separate `x_sq` tensor); doxygen + clang-format applied.

No P0/P1/P2 issues. Registration is collision-free: `"rms_norm"` (CPU/AppContext) and `"rmsnorm"` (OpenCL/ClContext) live in separate contexts with separate factory maps.

---

### 3. Latest-main Integration Stability Review

Clean integration after conflict resolution; `transformer.cpp` retains all upstream features and the Android build references the core library path (no dangling `../layers/rms_norm.cpp`). The full CI matrix is green.

---

### 📝 Overall Opinion ( Approve )

Clean promotion-to-core with full backward support, conflict and aliasing/guard issues resolved, and a green CI matrix. Approve.
